### PR TITLE
[FEATURE][processing] New dbscan spatial clustering algorithm

### PR DESCRIPTION
**Pull request by nyalldawson**
Implements an optimised DBSCAN density based scanning cluster approach for clustering 2d point features.

### CODE REVIEW

**Comment by nyalldawson** [comment]
Would love a code review here! @m-kuhn @pblottiere @wonder-sk?

File: src/analysis/processing/qgsalgorithmdbscanclustering.cpp [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+
+bool operator <( const QgsSpatialIndexKDBushData &data, const QgsFeatureId id )
+{
+  return data.id < id;
+};
+
+bool operator <( const QgsFeatureId id, const QgsSpatialIndexKDBushData &data )
+{
+  return id < data.id;
+};
```
Comment (wonder-sk) [inline src/analysis/processing/qgsalgorithmdbscanclustering.cpp]: Are the less than operators used anywhere? I guess they are not needed for unordered_set?

File: src/analysis/processing/qgsalgorithmdbscanclustering.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+#include "qgsprocessingalgorithm.h"
+#include <unordered_map>
+
+class QgsSpatialIndexKDBush;
+
+///@cond PRIVATE
+
+
+/**
+ * Native k-means clustering algorithm.
```
Comment (wonder-sk) [inline src/analysis/processing/qgsalgorithmdbscanclustering.h]: Copy/paste...

**Review by wonder-sk — APPROVED** [review]
Looking forward to have another useful algorithm in the toolbox! The code looks great.

I was thinking that in theory we could use KD-tree's internal data inside dbscan() instead of a using feature iterator, but most likely it is not worth the effort :-)

File: src/analysis/processing/qgsalgorithmdbscanclustering.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/analysis/processing/qgsalgorithmdbscanclustering.cpp]: I'll double check again. These may have only been required for an earlier iteration of this work.

**Comment by nyalldawson** [comment]
> I was thinking that in theory we could use KD-tree's internal data inside dbscan() instead of a using feature iterator, but most likely it is not worth the effort :-)

I'm gonna try anyway and see what speed difference it makes. Because.... Bragging rights for an awesome fast algorithm ;)

**Comment by nyalldawson** [comment]
Thanks for the review @wonder-sk!

File: src/analysis/processing/qgsalgorithmdbscanclustering.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/analysis/processing/qgsalgorithmdbscanclustering.cpp]: Thanks - this identified an issue in the code (was using std::set_difference, which shouldn't be used for unordered sets). Shaved about 25% off the processing time by fixing this :smile:

**Comment by nyalldawson** [comment]
> I was thinking that in theory we could use KD-tree's internal data inside dbscan() instead of a using feature iterator, but most likely it is not worth the effort :-)

Ok, negligible difference with this change. I'll leave the original approach because it allows for slightly better user feedback (errors reported when non-single point features are present)

File: src/analysis/processing/qgsalgorithmdbscanclustering.cpp (hunk 1 above)
Comment (wonder-sk) [reply src/analysis/processing/qgsalgorithmdbscanclustering.cpp]: Awesome!

**Pull request merged**